### PR TITLE
Fixed #27928 -- Avoided SET/DROP DEFAULT unless a field changes from null to non-null.

### PR DESCRIPTION
**Pull request by charettes**
[Ticket #27928](https://code.djangoproject.com/ticket/27928).

### CODE REVIEW

**Comment by charettes** [comment]
I also noticed that the following `SchemaTests` methods seem to be testing the exact same thing:

1. `test_alter_field_default_dropped`
2. `test_alter_null_to_not_null`
3. `test_alter_null_to_not_null_keeping_default`

**Comment by timgraham** [comment]
If we fixed SQLite not to do pointless operations (the usual create table, move data over, rename table), could we use `self.assertNumQueries(0):` for the test? I feel that would be less brittle than inspecting the SQL that runs.

**Comment by charettes** [comment]
@timgraham I also thought of the same thing but didn't spend any time investigating how much effort that would require. I'll give it a try and report back here.

**Comment by charettes** [comment]
After a bit of investigation I believe teaching the SQLite backend to avoid recreating tables when not necessary would require a non-trivial amount of work.

The only solution I could think of would be to make `_remake_table` compare the `CREATE TABLE`  of the actual table and to the one it should be rebuilt to in order to abort the process if definitions are equivalent.

The main issue is that there isn't any _build the CREATE TABLE statement for this model and return it_ method defined on schema editor; all the logic is embedded in `create_model` which performs the operation directly.

We _could_ theoretically create a new instance of `self.__class__` by passing `record=True` in order to gather the SQL by calling `create_model()` on it but this sounds hacky at best in a method that already does a lot of voodoo to achieve its goal.

**Comment by charettes** [comment]
@timgraham I just pushed a commit that demonstrates why preventing SQLite from remaking tables will take a bit more time.

1. We can't use `assertNumQueries(0)` because SQLite's `DatabaseSchemaEditor.__enter__`/`__exit__` executes uncollectable `PRAGMA foreign_keys` queries making the `self.__class__(collect_sql=True)` approach unusable.
2. Some tests are written in a way that abuses the `alter_field` API by passing a model which definition doesn't include the provided `old_field`. For example in [`schema.tests.SchemaTests.test_indexes`](https://github.com/django/django/blob/73c30b79c0c022464105b6031236571e2ff1d537/tests/schema/tests.py#L1681) we pass the following arguments

```python
class model(models.Model):
    title = models.CharField(max_length=100, db_index=True)

old_field = models.CharField(max_length=100, db_index=False)
new_field = models.CharField(max_length=100, db_index=True)
```

Where we should be passing a model definition with a `db_index=False` title instead. Really makes me wonder why the `alter_field()` API doesn't take have `(model, old_field_name, new_field)` signature to prevent these kind of abuses.

File: tests/schema/tests.py [hunk 1]
```diff
@@ -1956,6 +1956,25 @@ def test_alter_field_default_dropped(self):
...
+            editor.create_model(AuthorWithDefaultHeight)
+        old_field = AuthorWithDefaultHeight._meta.get_field('height')
+        new_default = old_field.default * 2
+        new_field = PositiveIntegerField(null=True, blank=True, default=new_default)
+        new_field.set_attributes_from_name('height')
+        with CaptureQueriesContext(connection) as ctx, connection.schema_editor() as editor:
+            editor.alter_field(AuthorWithDefaultHeight, old_field, new_field, strict=True)
+        add_default_sql = editor.sql_alter_column_default % {
+            'column': editor.quote_name(new_field.name),
+            'default': new_default
```
Comment (timgraham) [inline tests/schema/tests.py]: missing trailing comma

**Comment by charettes** [comment]
@timgraham makes sense, I doubt other backends naively rebuild tables on field alteration. Pushed a version that uses `skipIf` and `assertNumQueries(0)`.

File: tests/schema/tests.py [hunk 2]
```diff
@@ -1956,6 +1956,21 @@ def test_alter_field_default_dropped(self):
             if connection.features.can_introspect_default:
                 self.assertIsNone(field.default)
 
+    @unittest.skipIf(connection.vendor == 'sqlite', 'SQLite naively remake table on field alteration.')
+    def test_alter_field_default_doesnt_perfom_queries(self):
+        """
+        No queries should be performed if a field default changes unless a
```
Comment (timgraham) [inline tests/schema/tests.py]: I guess I'd say, "No queries are performed if a field default changes and the field's not changing from null to non-null." 
The "unless" phrasing may give the idea that case is also tested here.

**Pull request merged**